### PR TITLE
Merge factory objects into a single object

### PR DESCRIPTION
**Pull request by NathanW2**
Merge `QgsLayerStylingPanelFactory` and `QgsVectorLayerPropertiesPage` into a single `QgsMapLayerPanelFactory`.  As the APIs were pretty much the same it made sense to really merge the two. 

There was nothing really styling based about `QgsLayerStylingPanel` so it has been renamed to `QgsMapLayerPanel` to be more generic to fit both 

@m-kuhn @nyalldawson for your review.

**Note** This isn't going to pass the build, mainly just for API review.

### CODE REVIEW

File: src/gui/qgsmaplayerpropertiesfactory.h [hunk 1]
```diff
@@ -19,36 +19,58 @@
...
 
     /** Destructor */
-    virtual ~QgsMapLayerPropertiesFactory();
+    virtual ~QgsMapLayerPanelFactory();
+
+    /**
+     * @brief The icon that will be shown in the UI for the panel.
+     * @return A QIcon for the panel icon.
+     */
+    virtual QIcon icon() { return QIcon(); }
```
Comment (m-kuhn) [inline src/gui/qgsmaplayerpropertiesfactory.h]: virtual QIcon icon() const

File: src/gui/qgsmaplayerpropertiesfactory.h [hunk 2]
```diff
@@ -19,36 +19,58 @@
...
+     * @note This may or may not be shown to the user.
+     * @return Title of the panel
+     */
+    virtual QString title() { return QString(); }
+
+    /**
+     * @brief Check if the layer is supported for this widget.
+     * @return True if this layer is supported for this widget
+     */
+    virtual bool supportsLayer( QgsMapLayer *layer );
```
Comment (m-kuhn) [inline src/gui/qgsmaplayerpropertiesfactory.h]: Actually, `const` all of them

**Comment by m-kuhn** [comment]
Good job, thanks a lot!

I was wondering, is it also a *panel* when used on a modal dialog (properties dialog)? Or would `QgsLayerConfigWidget` be a better suited name?

**Comment by NathanW2** [comment]
I'm happy with `QgsLayerConfigWidget` as a name.

**Comment by m-kuhn** [comment]
I guess you are still working on it and some class names may change. Just make a grep at the end to check that all `PropertiesFactory` etc. are properly replace with `LayerConfigWidgetFactory` etc.

**Comment by NathanW2** [comment]
Yeah I'll clean up all the file names etc.

So we are good with with merge and rename?

**Comment by m-kuhn** [comment]
Rename is good for me. Can you wait with the merge? I'll have a look at it again once it's updated.

**Comment by NathanW2** [comment]
Yes I will wait until all good here

File: python/gui/qgsmaplayerconfigwidgetfactory.sip [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    virtual bool supportsLayer( QgsMapLayer *layer ) const;
+
+    /**
+     * @brief Create a new properties page
+     * @param layer The layer for which to create the page
+     * @param parent The parent widget
+     * @return The new properties page instance
+     */
+    virtual QgsMapLayerConfigWidget* createPropertiesPage( QgsVectorLayer* layer, QWidget* parent = nullptr ) const;
```
Comment (m-kuhn) [inline python/gui/qgsmaplayerconfigwidgetfactory.sip]: Proper sip annotations for`/Factory/`, `TransferThis`

**Comment by NathanW2** [comment]
Manual merged.

**Pull request closed without merging**